### PR TITLE
[Event Hubs Client] Idempotent Producer Support Scaffolding

### PR DESCRIPTION
**Pull request by jsquire**
# Summary

The focus of these changes is to implement the idempotent publishing feature scaffolding for the supporting types.  The `EventHubProducerClient` changes
are not included in this set, and no service operations are yet active.

**Note:**  These API changes are part of a preview feature.  The design and API surface were informally reviewed with the language architect but have not yet undergone formal review with the architecture board.  I expect some changes, particularly around naming.

# Last Upstream Rebase

Friday, August 14, 1:28pm (EDT)

# References and Related Issues

- [Idempotent Publishing Design](https://gist.github.com/jsquire/1cc4db6b3ca4ef13d26dc5315483555b)
- [Idempotent Publishing Requirements with Usage](https://gist.github.com/jsquire/0f3bc5701388fe03ccd027e86f1994f1)
- [Idempotent Publishing Support](https://github.com/Azure/azure-sdk-for-net/issues/13941) ([#13941](https://github.com/Azure/azure-sdk-for-net/issues/13941))

### CODE REVIEW

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Consumer/EventHubConsumerClient.cs [hunk 1]
```diff
@@ -315,7 +315,7 @@ public virtual async Task<string[]> GetPartitionIdsAsync(CancellationToken cance
         /// <returns>The set of information for the requested partition under the Event Hub this client is associated with.</returns>
         ///
         public virtual async Task<PartitionProperties> GetPartitionPropertiesAsync(string partitionId,
-                                                                             CancellationToken cancellationToken = default)
+                                                                                   CancellationToken cancellationToken = default)
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Consumer/EventHubConsumerClient.cs:318]: This is orthogonal;  it's just riding along because it was annoying me.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Consumer/ReadEventOptions.cs [hunk 2]
```diff
@@ -26,11 +26,11 @@ public class ReadEventOptions
         /// <summary>
         ///   When populated, the owner level indicates that a reading is intended to be performed exclusively for events in the
         ///   requested partition and for the associated consumer group.  To do so, reading will attempt to assert ownership
-        ///   over the partition; in the case where more than one exclusive reader attempts to assert ownership for the same
-        ///   partition/consumer group pair, the one having a larger <see cref="OwnerLevel"/> value will "win."
+        ///   over the partition; in the case where more than one exclusive reader in the consumer group attempts to assert ownership
+        ///   for the same partition, the one having a larger <see cref="OwnerLevel"/> value will "win".
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Consumer/ReadEventOptions.cs:30]: This is orthogonal;  it's just riding along because it was annoying me.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubsException.cs [hunk 3]
```diff
@@ -216,7 +217,13 @@ public enum FailureReason
             ServiceTimeout,
 
             /// <summary>There was a general communications error encountered when interacting with the Azure Event Hubs service.</summary>
-            ServiceCommunicationProblem
+            ServiceCommunicationProblem,
+
+            /// <summary>A client was forcefully disconnected from an Event Hub instance.  This typically occurs when another consumer with higher <see cref="PartitionPublishingOptions.OwnerLevel" /> asserts ownership over the partition and producer group.</summary>
+            ProducerDisconnected,
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubsException.cs:223]: There is an open item for discussion in the design around this member.  Including it for now.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/PartitionPublishingOptions.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+namespace Azure.Messaging.EventHubs.Producer
+{
+    /// <summary>
+    ///   The set of options that can be specified for an <see cref="EventHubProducerClient" />
+    ///   to influence its behavior when publishing directly to an Event Hub partition.
+    /// </summary>
+    ///
+    /// <remarks>
+    ///   These options are ignored  when publishing to the Event Hubs gateway for automatic
+    ///   routing nor when using a partition key.
```
Comment (christothes) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/PartitionPublishingOptions.cs:13]: same question about "nor" here

**Review by christothes — APPROVED** [review]
👍 Just a minor question.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubProducerClientOptions.cs [hunk 5]
```diff
@@ -20,6 +21,31 @@ public class EventHubProducerClientOptions
...
+        ///   The set of options that can be specified to influence publishing behavior specific to the configured Event Hub partition.  These
+        ///   options are not necessary in the majority of scenarios and are intended for use with specialized scenarios, such as when
+        ///   recovering the state used for idempotent publishing.
+        ///
+        ///   <para>It is highly recommended that these options only be specified if there is a proven need to do so; Incorrectly configuring these
+        ///   values may result in an <see cref="EventHubProducerClient" /> instance that is unable to publish to the Event Hubs.</para>
+        /// </summary>
+        ///
+        /// <remarks>
+        ///   These options are ignored when publishing to the Event Hubs gateway for automatic routing nor when using a partition key.
```
[the remark this answers is not in the document]
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubProducerClientOptions.cs:44]: ooops!    That should have been "or".    Thanks!

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/PartitionPublishingOptions.cs (hunk 4 above)
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/PartitionPublishingOptions.cs:13]: Ditto

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubsException.cs [hunk 6]
```diff
@@ -216,7 +217,13 @@ public enum FailureReason
...
+
+            /// <summary>A client is in an invalid state from which it cannot recover.  It is recommended that the client be closed and recreated to force reinitialization of state.</summary>
+            InvalidClientState
```
Comment (richardpark-msft) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubsException.cs:226]: So would this fall under a "this is recoverable, so the SDK should just do this internally"? Sort of like retries?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/PartitionPublishingProperties.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        /// </summary>
+        ///
+        public short? OwnerLevel { get; set; }
+
+        /// <summary>
+        ///   The sequence number assigned to the event that was most recently published to the associated partition
+        ///   successfully.
+        /// </summary>
+        ///
+        public int? LastPublishedSequenceNumber   { get; set; }
```
Comment (richardpark-msft) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/PartitionPublishingProperties.cs:37]: We kind of discussed this before - should we come up with a different name for this to avoid confusion/overlap with the SequenceNumber that most people are going to interact with?

Would it be bad to call it a Count, for instance? Since it has to be ordinal and can't contain gaps?
(just spitballing)

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubsException.cs (hunk 6 above)
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubsException.cs:226]: No; this is a sign that there is a deeper problem where the service and client disagree on the state of the world.  We discussed this quite a bit and determined that we cannot deterministically prove that it is safe to recover.  

As the comment indicates, this is a signal that the client is corrupt and callers should create a new instance.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/PartitionPublishingProperties.cs (hunk 7 above)
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/PartitionPublishingProperties.cs:37]: Naming is still in the "here's what the working group accepted during design reviews" stage.   I absolutely expect changes as we move into arch board reviews and such.

That said "count" here wouldn't be accurate, I don't think.  I don't believe that it adequately communicates the scope "count sent for this particular producer group and owner level combination which may have started at an arbitrary point."    While "sequence number" also lacks that level of insight, I think that it doesn't come with a strong a connotation that it starts at 0|1 and counts up.

Comment (richardpark-msft) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/PartitionPublishingProperties.cs:37]: Yes, naming is always the hardest part. 

Is the idea of having this be a "hidden" property still on the table? Or are we dead-set on exposing the value?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubsException.cs (hunk 6 above)
Comment (richardpark-msft) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubsException.cs:226]: Yeah, I read the comment and thought "yes, but....how corrupt?".

File: sdk/eventhub/Azure.Messaging.EventHubs/api/Azure.Messaging.EventHubs.netstandard2.0.cs [hunk 8]
```diff
@@ -10,6 +10,7 @@ public partial class EventData
         public long Offset { get { throw null; } }
         public string PartitionKey { get { throw null; } }
         public System.Collections.Generic.IDictionary<string, object> Properties { get { throw null; } }
+        public int? PublishedSequenceNumber { get { throw null; } }
```
Comment (YijunXieMS) [inline sdk/eventhub/Azure.Messaging.EventHubs/api/Azure.Messaging.EventHubs.netstandard2.0.cs:13]: Is this number transient in the service because it's only useful for a short period of time?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubProducerClientOptions.cs [hunk 9]
```diff
@@ -20,6 +21,31 @@ public class EventHubProducerClientOptions
...
+        ///
+        ///   <para>It is highly recommended that these options only be specified if there is a proven need to do so; Incorrectly configuring these
+        ///   values may result in an <see cref="EventHubProducerClient" /> instance that is unable to publish to the Event Hubs.</para>
+        /// </summary>
+        ///
+        /// <remarks>
+        ///   These options are ignored when publishing to the Event Hubs gateway for automatic routing or when using a partition key.
+        /// </remarks>
+        ///
+        public Dictionary<string, PartitionPublishingOptions> PartitionOptions { get; } = new Dictionary<string, PartitionPublishingOptions>();
```
Comment (YijunXieMS) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubProducerClientOptions.cs:47]: Is PartitionOptions too generic a name? If someday we have other options for every partition, this name sounds it has all options for partitions. But its type is PartitionPublishingOptions.

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubProducerClientOptions.cs:47]: These are attached to the producer client, so there's an implied context; I don't see value in making the property name more complicated.   As you noted, the type also provides an explicit context.
